### ISSUE TITLE
Make it easier to train composable rulesets

### ISSUE THREAD

**Issue by biancadanforth** [mozilla/fathom#176]
For the Smoot project, we are classifying pages by category. One of the categories is `"article"`, and the approach I took was to have two Fathom `type`s, `"paragraph"` and `"article"`, where an article is composed of paragraphs, and therefore most of the rules in the `"article"` ruleset are based on the output of the `"paragraph"` ruleset[1].

This has proven to be a very effective approach in terms of accurately identifying article pages, but the process of developing and iterating on the ruleset is quite tedious:

Right now, if I want to test a new rule for the `"paragraph"` ruleset I have to:
1. Vectorize the paragraph ruleset in FathomFox.
2. Train the paragraph ruleset using `fathom-train`.
3. Plug in the coefficients and bias from training back into the ruleset.
4. Vectorize the article ruleset in FathomFox.
5. Train the article ruleset using `fathom-train`.

And this is just for one set of samples (i.e. a training set). If I also have a validation and/or test set, then each of those sets must also be separately vectorized.

Building on issue #175 , it would be great if any `type` dependencies could be specified and these steps automated.

[1]:
```js
function makeRuleset(coeffs, biases) {
  return ruleset([
    /**
      * Paragraph rules
    */
    rule(dom("p, pre").when(isElementVisible), type("paragraph")),
    rule(dom("div").when(isElementVisible).when(divHasBrChildElement), type("paragraph").note(numParagraphTextNodesInDiv)),
    rule(dom("div").when(isElementVisible).when(divHasOnlyTextNodesAnchorElementsOrSpanElements), type("paragraph").note(numParagraphTextNodesInDiv)),
    rule(type("paragraph"), score(pElementHasListItemAncestor), {name: "pElementHasListItemAncestor"}),
    rule(type("paragraph"), score(hasLongTextContent), {name: "hasLongTextContent"}),
    rule(type("paragraph"), score(containsElipsisAtEndOfText), {name: "containsElipsisAtEndOfText"}),
    rule(type("paragraph"), score(classNameIncludesCaption), {name: "classNameIncludesCaption"}),
    rule(type("paragraph").max(), out("paragraph")),

    /**
      * Article rules
    */
    rule(dom("html"), type("article")),
    rule(type("article"), score(hasEnoughParagraphs), {name: "hasEnoughParagraphs"}),
    rule(type("article"), score(hasExactlyOneArticleElement), {name: "hasExactlyOneArticleElement"}),
    rule(type("article"), score(paragraphElementsHaveSiblingsWithSameTagName), {name: "paragraphElementsHaveSiblingsWithSameTagName"}),
    rule(type("article"), score(mostParagraphElementsAreHorizontallyAligned), {name: "mostParagraphElementsAreHorizontallyAligned"}),
    rule(type("article"), score(moreParagraphElementsThanListItemsOrTableRows), {name: "moreParagraphElementsThanListItemsOrTableRows"}),
    rule(type("article"), score(headerElementIsSiblingToParagraphElements), {name: "headerElementIsSiblingToParagraphElements"}),
    rule(type("article"), score(hasMultipleArticleElements), {name: "hasMultipleArticleElements"}),
    rule(type("article"), score(hasMultipleParagraphsWhoseClassNameIncludesArticle), {name: "hasMultipleParagraphsWhoseClassNameIncludesArticle"}),
    rule(type("article"), out("article"))
  ],
  coeffs,
  biases);
}
```
